### PR TITLE
Adjust claim_outputs() for very small inputs

### LINKED ISSUE #1373: Check storage deposit minimum
Check all places where we might assume that the amount is >= min storage deposit, but it might actually not  (for example output consolidation/claiming (the get_basic_outputs_for_additional_inputs() there for example)

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Adjust claim_outputs() to work even if an input has less amount than min storage deposit, because before the code assumed that if there is an input, that it can be used to create a new output which might not be the case anymore

## Links to any relevant issues

https://github.com/iotaledger/iota-sdk/issues/1373

## Type of change

- Bug fix (a non-breaking change which fixes an issue)

## How the change has been tested

Didn't test, there isn't really a way without custom snapshot creation

### CODE REVIEW

File: sdk/CHANGELOG.md [hunk 1]
```diff
@@ -24,6 +24,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - Update protocol params and addresses with correct bech32 HRP in `Wallet::set_client_options()`;
+- `Account::claim_outputs()` if an input have less amount than min storage deposit;
```
Comment (thibault-martinez) [inline sdk/CHANGELOG.md]:
```suggestion
- `Account::claim_outputs()` if an input has less amount than min storage deposit;
```

**Review by DaughterOfMars — APPROVED** [review]

File: sdk/src/wallet/account/operations/output_claiming.rs [hunk 2]
```diff
@@ -266,7 +275,7 @@ where
                 // build new output with same amount, nft_id, immutable/feature blocks and native tokens, just
                 // updated address unlock conditions
 
-                let nft_output = if possible_additional_inputs.is_empty() {
+                let nft_output = if !enough_amount_for_basic_output {
                     // Only update address and nft id if we have no additional inputs which can provide the storage
```
Comment (thibault-martinez) [inline sdk/src/wallet/account/operations/output_claiming.rs:279]: I don't get the comment, address and nft ID are in both if and else ?

File: sdk/src/wallet/account/operations/output_claiming.rs [hunk 3]
```diff
@@ -266,7 +275,7 @@ where
...
                     // deposit for the remaining amount and possible NTs
```
Comment (thibault-martinez) [inline sdk/src/wallet/account/operations/output_claiming.rs:280]: I don't understand this comment, we do same in both if/else?

Comment (Thoralf-M) [reply sdk/src/wallet/account/operations/output_claiming.rs:280]: It's not `only` in this case, but `only` do this and don't update anything else

File: sdk/src/wallet/account/operations/output_claiming.rs [hunk 4]
```diff
@@ -242,6 +242,15 @@ where
...
         let mut required_amount_for_nfts = 0;
         let mut new_native_tokens = NativeTokensBuilder::new();
+
+        // There can be outputs with less amount than min required storage deposit, so we have to check that we
+        // have enough amount to create a new basic output
+        let enough_amount_for_basic_output = possible_additional_inputs
+            .iter()
+            .map(|i| i.output.amount())
+            .sum::<u64>()
+            >= MinimumStorageDepositBasicOutput::new(rent_structure, token_supply).finish()?;
```
Comment (kwek20) [inline sdk/src/wallet/account/operations/output_claiming.rs:252]: wouldnt this just be Output::AMOUNT_MIN always?

Comment (DaughterOfMars) [reply sdk/src/wallet/account/operations/output_claiming.rs:252]: No? `Output::AMOUNT_MIN == 1`

Comment (kwek20) [reply sdk/src/wallet/account/operations/output_claiming.rs:252]: tru, i derped

**Review by kwek20 — DISMISSED** [review]

**Review by DaughterOfMars — DISMISSED** [review]

File: sdk/src/wallet/account/operations/output_claiming.rs [hunk 5]
```diff
@@ -242,6 +242,15 @@ where
         let mut available_amount = 0;
         let mut required_amount_for_nfts = 0;
         let mut new_native_tokens = NativeTokensBuilder::new();
+
+        // There can be outputs with less amount than min required storage deposit, so we have to check that we
+        // have enough amount to create a new basic output
+        let enough_amount_for_basic_output = possible_additional_inputs
+            .iter()
+            .map(|i| i.output.amount())
```
Comment (thibault-martinez) [inline sdk/src/wallet/account/operations/output_claiming.rs:250]: What if there are returns in the possible_additional_inputs and in the end the actual sum is not enough ?

Comment (kwek20) [reply sdk/src/wallet/account/operations/output_claiming.rs:250]: I checked the code, and these additional_inputs are unlock_conditions().len() == 1, so allegedly only AddressUnlockCondition outputs. We may want to... foolproof that check :)

Comment (Thoralf-M) [reply sdk/src/wallet/account/operations/output_claiming.rs:250]: It's also documented for `get_basic_outputs_for_additional_inputs()` that there will only be the address unlock condition, exactly for this reason that we don't have to take care of sdruc or check timelock expiration again

**Review by thibault-martinez — COMMENTED** [review]
Conflicts, this should have been merged before 1.1.1 :(

File: sdk/CHANGELOG.md [hunk 6]
```diff
@@ -19,6 +19,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Security -->
 
+## 1.1.2 - 2023-MM-DD
+### Fixed
```
Comment (thibault-martinez) [inline sdk/CHANGELOG.md:24]:
```suggestion
### Fixed

```

**Review by DaughterOfMars — DISMISSED** [review]

File: sdk/CHANGELOG.md (hunk 6 above)
Comment (thibault-martinez) [reply sdk/CHANGELOG.md:24]: Why resolved ?

Comment (thibault-martinez) [reply sdk/CHANGELOG.md:24]: Oh damn I fucked up the suggestion sorry

Comment (thibault-martinez) [reply sdk/CHANGELOG.md:24]: Can you fix the spacing stuff pls

Comment (Thoralf-M) [reply sdk/CHANGELOG.md:24]: Done

**Review by DaughterOfMars — DISMISSED** [review]

File: sdk/CHANGELOG.md [hunk 7]
```diff
@@ -19,6 +19,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Security -->
 
+## 1.1.2 - 2023-MM-DD
+
+### Fixed
+
+- `Account::claim_outputs()` if an input have less amount than min storage deposit;
```
Comment (thibault-martinez) [inline sdk/CHANGELOG.md]:
```suggestion
- `Account::claim_outputs()` if an input has less amount than min storage deposit;
```

**Review by thibault-martinez — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]

**Pull request merged**